pghoard/compressor.py

```diff
@@ -4,23 +4,87 @@
 Copyright (c) 2016 Ohmu Ltd
 See LICENSE for details
 """
+import enum
 import hashlib
 import logging
 import math
 import os
 import socket
 import time
+from collections import defaultdict
+from dataclasses import dataclass
 from io import BytesIO
+from pathlib import Path
 from queue import Empty, Queue
 from tempfile import NamedTemporaryFile
 from threading import Event, Thread
-from typing import Dict, Optional, Set
+from typing import Any, BinaryIO, Dict, Optional, Set, Union
 
 from pghoard import config as pgh_config
 from pghoard import wal
-from pghoard.common import write_json_file
+from pghoard.common import (CallbackEvent, CallbackQueue, FileType, FileTypePrefixes, QuitEvent, StrEnum, write_json_file)
 from pghoard.metrics import Metrics
-from pghoard.rohmu import errors, rohmufile
+from pghoard.rohmu import rohmufile
+from pghoard.transfer import TransferQueue, UploadEvent
+
+
+@enum.unique
+class CompressionOperation(StrEnum):
+    Decompress = "DECOMPRESS"
+    Compress = "COMPRESS"
+
+
+@dataclass(frozen=True)
+class BaseCompressorEvent:
+    file_type: FileType
+    file_path: Path
+    backup_site_key: str
+    source_data: Union[BinaryIO, Path]
+    callback_queue: CallbackQueue
+    metadata: Dict[str, str]
+
+
+@dataclass(frozen=True)
+class _BaseCompressorEventDefaults:
+    file_size: Optional[int] = None
+    opaque: Optional[Any] = None
+
+
+@dataclass(frozen=True)
+class CompressionEvent(BaseCompressorEvent):
+    delete_file_after_compression: bool = False
+    compress_to_memory: bool = False
+
+    @property
+    def operation(self):
+        return CompressionOperation.Compress
+
+
+@dataclass(frozen=True)
+class DecompressionEvent(BaseCompressorEvent):
+
+    destination_path: Path
+
+    @property
+    def operation(self):
+        return CompressionOperation.Decompress
+
+
+# Should be changed to Queue[CompressionEvent] once
```

Comment:
nit.. Since `QuitEvent` is separate, comment should say `Queue[Union[CompressionEvent, Literal[QuitEvent]]]`